src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,272 @@
+        {
+            lock (_gate)
+            {
+                if (_registered)
+                {
+                    _handler(context);
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        private static int OnPosixSignal(int signo, PosixSignal signal)
+        {
+            PosixSignalRegistration?[]? registrations = GetRegistrations(signo);
+            if (registrations != null)
+            {
+                // This is called on the native signal handling thread. We need to move to another thread so
+                // signal handling is not blocked. Otherwise we may get deadlocked when the handler depends
+                // on work triggered from the signal handling thread.
+
+                // For terminate/interrupt signals we use a dedicated Thread
+                // in case the ThreadPool is saturated.
+                bool useDedicatedThread = signal == PosixSignal.SIGINT ||
+                                            signal == PosixSignal.SIGQUIT ||
+                                            signal == PosixSignal.SIGTERM;
+                if (useDedicatedThread)
+                {
+                    Thread handlerThread = new Thread(HandleSignal)
+                    {
+                        IsBackground = true,
+                        Name = ".NET Signal Handler"
+                    };
+                    handlerThread.UnsafeStart((signo, registrations));
+                }
+                else
+                {
+                    ThreadPool.UnsafeQueueUserWorkItem(HandleSignal, (signo, registrations));
+                }
+                return 1;
+            }
+            return 0;
+        }
+
+        private static PosixSignalRegistration?[]? GetRegistrations(int signo)
+        {
+            lock (s_registrations)
+            {
+                if (s_registrations.TryGetValue(signo, out List<WeakReference<PosixSignalRegistration>>? signalRegistrations))
+                {
+                    if (signalRegistrations.Count != 0)
+                    {
+                        var registrations = new PosixSignalRegistration?[signalRegistrations.Count];
+                        bool hasRegistrations = false;
+                        for (int i = 0; i < signalRegistrations.Count; i++)
+                        {
+                            if (signalRegistrations[i].TryGetTarget(out PosixSignalRegistration? registration))
+                            {
+                                registrations[i] = registration;
+                                hasRegistrations = true;
+                            }
+                            else
+                            {
+                                // WeakReference no longer holds an object. PosixSignalRegistration got finalized.
+                                signalRegistrations.RemoveAt(i);
```

Comment:
This isn't ideal... worst case it could be O(N^2).  Instead, you can do something like `signalRegistrations[i] = null`, and then after the loop, do `signalRegistrations.RemoveAll(item => item is null);`, or something like that.